src/com/ms/silverking/time/StopwatchBasedTimer.java

```diff
@@ -10,197 +10,197 @@
-    public double getTimeLimitSeconds() {
-        return nanos2seconds( getTimeLimitNanos() );
-    }
-
-    @Override
-    public BigDecimal getTimeLimitSecondsBD() {
-        return nanos2secondsBD( getTimeLimitNanos() );
-    }
-    
-    // StopwatchWrapper methods
-
-    @Override
-    public void start() {
-        sw.start();
-    }
-
-    @Override
-    public void stop() {
-        sw.stop();
-    }
-
-    @Override
-    public void reset() {
-        sw.reset();
-    }
-
-    @Override
-    public long getElapsedNanos() {
-        return sw.getElapsedNanos();
-    }
-
-    @Override
-    public long getElapsedMillisLong() {
-        return sw.getElapsedMillisLong();
-    }
-
-    @Override
-    public int getElapsedMillis() {
-        return sw.getElapsedMillis();
-    }
-
-    @Override
-    public double getElapsedSeconds() {
-        return sw.getElapsedSeconds();
-    }
-
-    @Override
-    public BigDecimal getElapsedSecondsBD() {
-        return sw.getElapsedSecondsBD();
-    }
-
-    @Override
-    public long getSplitNanos() {
-        return sw.getSplitNanos();
-    }
-
-    @Override
-    public long getSplitMillisLong() {
-        return sw.getSplitMillisLong();
-    }
-
-    @Override
-    public int getSplitMillis() {
-        return sw.getSplitMillis();
-    }
-
-    @Override
-    public double getSplitSeconds() {
-        return sw.getSplitSeconds();
-    }
-
-    @Override
-    public BigDecimal getSplitSecondsBD() {
-        return sw.getSplitSecondsBD();
-    }
-
-    @Override
-    public String getName() {
-        return sw.getName();
-    }
-
-    @Override
-    public State getState() {
-        return sw.getState();
-    }
-
-    @Override
-    public boolean isRunning() {
-        return sw.isRunning();
-    }
-
-    @Override
-    public boolean isStopped() {
-        return sw.isStopped();
-    }
-
-    @Override
-    public String toStringElapsed() {
-        return sw.toStringElapsed() +":"+ getTimeLimitSeconds();
-    }
-
-    @Override
-    public String toStringSplit() {
-        return sw.toStringSplit() +":"+ getTimeLimitSeconds();
-    }
-    
-    @Override
-    public String toString() {
-        return toStringSplit();
-    }
-
-    @Override
-    public boolean await(Condition cv) throws InterruptedException {
-        return cv.await(getRemainingMillisLong(), TimeUnit.MILLISECONDS);
-    }
+  private final Stopwatch sw;
+  private final long limitNanos;
+
+  private StopwatchBasedTimer(Stopwatch sw, long limitNanos) {
+    this.sw = sw;
+    this.limitNanos = limitNanos;
+  }
+
+  public StopwatchBasedTimer(Stopwatch sw, TimeUnit unit, long limit) {
+    this(sw, TimeUnit.NANOSECONDS.convert(limit, unit));
+  }
+
+  @Override
+  public boolean hasExpired() {
+    return getRemainingNanos() <= 0;
+  }
+
+  @Override
+  public void waitForExpiration() {
+    while (!hasExpired()) {
+      try {
+        Thread.sleep(getRemainingMillisLong());
+      } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)